redbean: Add LuaCrypto compatible functions (plus some auxiliary functions)

mterron opened the pull request:
Implements part of Issue #1136

mterron committed:
Add LuaCrypto compatible functions (plus some auxiliary functions) as per #1136

mterron committed:
Add LuaCrypto compatible functions (plus some auxiliary functions) as per #1136

pkulchenko reviewed (commented):
Thank you for submitting this; looks great! Couple of suggestions on the API:

- I'd consider the following renaming for consistency: `csrGenerate` to `generateCsr` (or `createCsr`), `generatekeypair` to `generateKeyPair` and `PemToJwk` to `convertPemToJwk`.
- I wonder if `generatekeypair` can make its first parameter optional. I'd still allow it to be passed, but given that RSA has an optional integer parameter, if nothing is passed or an integer is passed, I'd run RSA key pair generation.
- why is luaopen_lcrypto there if it doesn't seem to be defined/used?

mterron committed:
Align function naming

mterron committed:
Make key type optional in crypto.generateKeyPair. Defaults to rsa

mterron commented on tool/net/lcrypto.c:
```diff
@@ -0,0 +1,1147 @@
+static int LuaRSAGenerateKeyPair(lua_State *L) {
+  char *private_key, *public_key;
+  size_t private_len, public_len;
+  int key_length = 2048;  // Default RSA key length
+
+  // Get key length from Lua (optional parameter)
+  if (lua_gettop(L) >= 1 && !lua_isnil(L, 1)) {
+    key_length = luaL_checkinteger(L, 1);
```
idea let me try that.

mterron committed:
Fix tests

mterron committed:
Remove old reference

mterron committed:
PROPERLY make arguments optional in crypto.generatekeypair thanks to @pkulchenko

pkulchenko reviewed (approved):
Looks good to me. Thanks!

@jart, does the API look good to you?

mterron committed:
Add AES based encryption and decryption in multiple modes (CBC, CTR & GCM)
Improve test coverage

mterron committed:
Unconditionally compile AES
Cleanup language on the test file

mterron committed:
Improve tests
Add definitions
Align function name

mterron committed:
Quiet tests

mterron committed:
Change of API for Encrypt and Decrypt.
The options are now passed in a table instead of positional parameters. This is not LuaCrypto compatible but it is a nicer interface.

mterron committed:
Merge parse aes parse options functions

pkulchenko reviewed (approved):
Had a chance to review added methods and updated API. Looks good to me. Thanks!

mterron committed:
Use built-in entropy generator and remove dependency on GetHardRandom
Add correct jwltopem and pemtojwk function
Expand tests

solisoft commented:
Awesome !

mterron committed:
Add RSA PSS support
Improve error messages
Improve parameter validation
Correct base64url encoding for JWK
Add support for optional claims to convertPemToJwk
Expand test coverage
Add basic definitions

pkulchenko commented:
@mterron, thank you for the updates! I liked added error checking and new functions! Noticed a couple of things to check on (couldn't leave comments on individual lines for some reason, so summarizing it here):
- line 502: Previously this would return NULL for an unsupported hash algorithm, but now it will store -1, which will then gets passed to `mbedtls_md_info_from_type` and other functions. Is that expected?
- similarly line 612, 723, 886; I'd consider implementing something similar to what you did on line 932.
- line 694: should hash_name be checked for validity?
- 808 and 810 have `DEBUGF` for the hash, but not other places with similar processing.
- I'm curious, why 989, 994 and 999 lines were replaced? should be the same result, no?
- `if` checks on lines 1369 and 1371 are the same, which makes the content of `else if` on line 1378 unreachable. Is there an issue with the logic?
- Indentation on line 1386 seems to be off. Also on line 1918.

mterron committed:
Address Paul's comments :)

mterron committed:
Cleanup hash validation
Improve JWK functions
Enable PKCS1 v2.1 support in MBEDTLS

mterron committed:
All modes available on "tiny" builds

reneleonhardt commented:
@pkulchenko After approval, why wasn't it merged?

jart commented:
Terror isn’t a good fit for the project. We need to find someone else
willing to own responsibility for this code.

On Mon, Jul 13, 2026 at 2:01 AM Rene Leonhardt ***@***.***>
wrote:

> *reneleonhardt* left a comment (jart/cosmopolitan#1422)
> <https://github.com/jart/cosmopolitan/pull/1422#issuecomment-4954892971>
>
> @pkulchenko <https://github.com/pkulchenko> After approval, why wasn't it
> merged?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jart/cosmopolitan/pull/1422?email_source=notifications&email_token=AAAMA3SPCLXKHNNBQXQMJQ35ER3NDA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJVGQ4DSMRZG4Y2M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4954892971>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMA3UWQLEIYENU4XDZM6D5ER3NDAVCNFSNUABFKJSXA33TNF2G64TZHMZDOMRUGU3TMMBWHNEXG43VMU5TGMJQGY4TKMBWGM22C5QC>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
